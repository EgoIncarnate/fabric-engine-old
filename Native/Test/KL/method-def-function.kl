struct Vec3
{
  Scalar x;
  Scalar y;
  Scalar z;
};

function Vec3 ( Scalar x, Scalar y, Scalar z )
{
  self.x = x;
  self.y = y;
  self.z = z;
}

function Scalar Vec3.normSq()
{
  return self.x*self.x + self.y*self.y + self.z*self.z;
}

function entry()
{
  var Vec3 vec3( 1.0, 2.0, 3.0 );
  report vec3.normSq();
}
