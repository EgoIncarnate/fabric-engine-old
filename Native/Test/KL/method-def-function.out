GlobalList
  StructDecl( 'Vec3' )
    MemberDecl( 'x', Scalar )
    MemberDecl( 'y', Scalar )
    MemberDecl( 'z', Scalar )
  Function( '' ('__construct_Vec3__Scalar__Scalar__Scalar'), void )
    ParamList
      Param( 'self', Vec3, l-value )
      Param( 'x', Scalar, r-value )
      Param( 'y', Scalar, r-value )
      Param( 'z', Scalar, r-value )
    CompoundStatement
      StatementList
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'x' )
              Var('self')
            Var('x')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'y' )
              Var('self')
            Var('y')
        ExprStatement
          AssignOp(ASSIGN_OP)
            StructMemberOp( 'z' )
              Var('self')
            Var('z')
  Function( '' ('__method_normSq__Vec3'), Scalar:r-value )
    ParamList
      Param( 'self', Vec3, r-value )
    CompoundStatement
      StatementList
        ReturnStatement
          BinOp( ADD )
            BinOp( ADD )
              BinOp( MUL )
                StructMemberOp( 'x' )
                  Var('self')
                StructMemberOp( 'x' )
                  Var('self')
              BinOp( MUL )
                StructMemberOp( 'y' )
                  Var('self')
                StructMemberOp( 'y' )
                  Var('self')
            BinOp( MUL )
              StructMemberOp( 'z' )
                Var('self')
              StructMemberOp( 'z' )
                Var('self')
  Function( 'entry' ('entry'), void )
    ParamList
    CompoundStatement
      StatementList
        StatementList
          InitializedVarDecl( VarDecl( 'vec3', Vec3 ), ArgList )
            ArgList
              ConstScalar( 1 )
              ConstScalar( 2 )
              ConstScalar( 3 )
        Report
          MethodOp( 'normSq' )
            Var('vec3')
            ArgList
14
